data/state/journalEntries.go

```diff
@@ -1,34 +1,120 @@
 package state
 
 import (
+	"bytes"
 	"fmt"
 
 	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/marshal"
 )
 
 type journalEntryCode struct {
-	codeHash []byte
-	updater  Updater
+	oldCodeEntry    *CodeEntry
+	oldCodeHash     []byte
+	newCodeHash     []byte
+	codeForEviction map[string]struct{}
+	newCode         map[string]struct{}
+	trie            Updater
+	marshalizer     marshal.Marshalizer
 }
 
 // NewJournalEntryCode creates a new instance of JournalEntryCode
-func NewJournalEntryCode(codeHash []byte, updater Updater) (*journalEntryCode, error) {
-	if check.IfNil(updater) {
+func NewJournalEntryCode(
+	oldCodeEntry *CodeEntry,
+	oldCodeHash []byte,
+	newCodeHash []byte,
+	codeForEviction map[string]struct{},
+	newCode map[string]struct{},
+	trie Updater,
+	marshalizer marshal.Marshalizer,
+) (*journalEntryCode, error) {
+	if check.IfNil(trie) {
 		return nil, ErrNilUpdater
 	}
-	if len(codeHash) == 0 {
-		return nil, ErrInvalidHash
+	if check.IfNil(marshalizer) {
+		return nil, ErrNilMarshalizer
+	}
+	if codeForEviction == nil {
+		return nil, ErrNilCodeForEvictionMap
+	}
+	if newCode == nil {
+		return nil, ErrNilNewCodeMap
 	}
 
 	return &journalEntryCode{
-		codeHash: codeHash,
-		updater:  updater,
+		oldCodeEntry:    oldCodeEntry,
+		oldCodeHash:     oldCodeHash,
+		newCodeHash:     newCodeHash,
+		trie:            trie,
+		marshalizer:     marshalizer,
+		codeForEviction: codeForEviction,
+		newCode:         newCode,
 	}, nil
 }
 
 // Revert applies undo operation
 func (jea *journalEntryCode) Revert() (AccountHandler, error) {
-	return nil, jea.updater.Update(jea.codeHash, nil)
+	if bytes.Equal(jea.oldCodeHash, jea.newCodeHash) {
+		return nil, nil
+	}
+
+	err := jea.revertOldCodeEntry()
+	if err != nil {
+		return nil, err
+	}
+
+	err = jea.revertNewCodeEntry()
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (jea *journalEntryCode) revertOldCodeEntry() error {
+	if len(jea.oldCodeHash) == 0 {
+		return nil
+	}
+
+	if jea.oldCodeEntry.NumReferences == 1 {
+		delete(jea.codeForEviction, string(jea.oldCodeHash))
+	}
+
```

Comment:
jea.oldCodeEntry.NumReferences-- ?

Reply:
+1

Reply:
No, because oldCodeEntry already has the numReferences set as before the changes. You can see in the accountDB that updateOldCodeEntry() returns unmodifiedOldCodeEntry that is used when creating a new journal entry.

Reply:
ok

Reply:
ok